fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -359,30 +335,56 @@ public void close() {
         return ret;
     }
 
-    /**
-     * Run after a single transactional call within runAsync.
-     */
+    @FunctionalInterface
     @VisibleForTesting
-    <R> Pair<R, Throwable> runAsyncPostTransaction(@Nullable R result, @Nullable Throwable exception,
-                                                   @Nullable AtomicLong recordsScanned) {
-        if (increaseLimitAfter > 0) {
-            if (exception == null) {
-                successCount++;
-                if (successCount >= increaseLimitAfter && limit < maxLimit) {
-                    increaseLimit();
-                }
-            } else {
-                successCount = 0;
-            }
-        }
-        if (recordsScanned != null) {
-            if (exception == null) {
-                totalRecordsScanned.addAndGet(recordsScanned.get());
+    interface ExceptionHandler<R> {
+        CompletableFuture<Boolean> apply(CompletableFuture<R> ret, AtomicLong toWait, Throwable e, @Nullable List<Object> additionalLogMessageKeyValues);
+    }
+
+    private <R> CompletableFuture<Boolean> completeExceptionally(CompletableFuture<R> ret, Throwable e) {
+        ret.completeExceptionally(runner.getDatabase().mapAsyncToSyncException(e));
+        return AsyncUtil.READY_FALSE;
+    }
+
+    @VisibleForTesting
+    <R> CompletableFuture<R> buildAsync(@Nonnull BiFunction<FDBRecordStore, AtomicLong, CompletableFuture<R>> buildFunction,
+                                        boolean limitControl,
+                                        @Nullable List<Object> additionalLogMessageKeyValues) {
+        AtomicLong recordsScanned = new AtomicLong(0);
+        return runAsync(store -> buildFunction.apply(store, recordsScanned),
+                // Run after a single transactional call within runAsync.
+                (result, exception) -> {
+                    if (limitControl) {
+                        tryToIncreaseLimit(exception);
+                    }
+                    // Update records scanned.
+                    if (exception == null) {
+                        totalRecordsScanned.addAndGet(recordsScanned.get());
+                    } else {
+                        recordsScanned.set(0);
+                    }
+                    return Pair.of(result, exception);
+                },
+                limitControl ? this::tryToDecreaseLimit : null,
+                additionalLogMessageKeyValues
+        );
+    }
+
+    @VisibleForTesting
+    <R> CompletableFuture<Boolean> tryToDecreaseLimit(CompletableFuture<R> ret, AtomicLong toWait, Throwable e, @Nullable List<Object> additionalLogMessageKeyValues) {
+        FDBException fdbE = getFDBException(e);
+        if (fdbE == null) {
+            return completeExceptionally(ret, e);
```

Comment:
Does it make sense to add the `additionalLogMessageKeyValues` values to the exception (if it is a CloudDBException)?  Maybe not...

Reply:
Yes, if `e` is a `LoggableException` we should add the details to that.